### ISSUE TITLE
Suggestion: Enforce exact include/exclude matching to merge loader configs

### ISSUE THREAD

**Issue by rockmacaca** [survivejs/webpack-merge#43]
Here's a simplified example to illustrate my problem:

Common config:
```javascript
rules: [{
  test: /\.scss$/
  include: [/node_modules/],
  use: [ExtractTextPlugin.loader(), 'css-loader', 'sass-loader']
}, { // everything else
  test: /\.scss$/
  use: ['css-loader', 'postcss-loader', 'sass-loader']
}]
```

Dev config:
```javascript
// strategy 'prepend'
rules: [{
  test: /\.scss$/
  use: ['style-loader']
}]
```

Prod config:
```javascript
// strategy 'prepend'
rules: [{
  test: /\.scss$/
  use: [ExtractTextPlugin.loader()]
}]
```

Since the use entry doesn't have an `include`/`exclude` field to match against, it merges with the first config with a matching `test`, the dev config results in
```javascript
rules: [{
  test: /\.scss$/
  include: [/node_modules/],
  use: ['style-loader', ExtractTextPlugin.loader(), 'css-loader', 'sass-loader']
}, { // everything else
  test: /\.scss$/
  use: ['css-loader', 'postcss-loader', 'sass-loader']
}]
```

I can't think of a good use case where `include` or `exclude` is applied afterward. Intuitively, they seem to be tied to the identity of a rule as much as the `test` property is. Yes, it isn't as DRY, but if the configs are stored in their own file it makes it very clear what rule is being overridden.

EDIT:
Apparently the example above doesn't actually work for the config I'm trying to make, but I think it's still useful. I am only just now realizing webpack appends when multiple loaders match the criteria!

**Comment by bebraw** [comment]
What I do is to not inherit styling configuration at all. It's easier to reason then even though it's a bit more configuration to write. Basically I have configuration "parts" that encapsulate the idea and then you can merge that so you get something like `setupCSS([... paths ...])`.

I guess in your case it would be enough to match the case not having `include` so you would get `use: ['style-loader', 'css-loader', 'postcss-loader', 'sass-loader']` over the current setup.

You could prepare a PR and we could go from there.

**Comment by bebraw** [comment]
@rockmacaca Do you want to proceed with this? Any thoughts?

**Comment by rockmacaca** [comment]
Sorry I haven't had time in the past couple weeks, I'm one of those people that gets busy with finals around this time of year ; )

I've still been using my own branch without any problems, I'll rebase onto 2.0 and add a test case and we'll see about merging.

**Comment by bebraw** [comment]
Ok, cool. Thanks for the heads up.

**Comment by rockmacaca** [comment]
It turns out I sort of worked around the issue when I changed some things a few weeks ago. I actually only needed to make sure that rules aren't merged where none of `test`, `include`, nor `exclude` are defined. This avoided introducing any breaking behavior, so for my own purposes, this is resolved with #44.

**Comment by bebraw** [comment]
Ok, cool. I'll try to review and merge soon so we get the improvements in.

If you could comment on #45, that would be great. I'm thinking the last scheme I discuss could work out nicely.
